src/sss_client/ssh/sss_ssh_knownhosts.c

```diff
@@ -0,0 +1,211 @@
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#define _GNU_SOURCE    /* For basename(3) */
+#include <string.h>
+#include <stdio.h>
+#include <talloc.h>
+#include <popt.h>
+#include <signal.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <netdb.h>
+
+#include "util/util.h"
+#include "util/crypto/sss_crypto.h"
+#include "util/sss_ssh.h"
+#include "sss_client/sss_cli.h"
+#include "sss_client/ssh/sss_ssh_client.h"
+
+static errno_t known_hosts(TALLOC_CTX *mem_ctx, const char *domain, const char *host,
+                           struct sss_ssh_ent **_ent, const char **_keyhost)
+{
+    TALLOC_CTX *tmp_ctx;
+    errno_t ret;
+    struct addrinfo ai_hint;
+    struct addrinfo *ai = NULL;
+    char canonhost[NI_MAXHOST];
+    const char *canonname = NULL;
+    const char *keyhost;
+    struct sss_ssh_ent *ent = NULL;
+
+    if (_ent == NULL || _keyhost == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Invalid NULL received\n");
+        ERROR("Internal error\n");
+        return EINVAL;
+    }
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to allocate tmp_ctx\n");
+        ERROR("Not enough memory\n");
+        return ENOMEM;
+    }
+
+    memset(&ai_hint, 0, sizeof(struct addrinfo));
+    ai_hint.ai_family = AF_UNSPEC;
+    ai_hint.ai_socktype = SOCK_STREAM;
+    ai_hint.ai_protocol = IPPROTO_TCP;
+    ai_hint.ai_flags = AI_NUMERICHOST | AI_NUMERICSERV;
+
+    DEBUG(SSSDBG_FUNC_DATA, "Looking up canonical name for: %s\n", host);
+    ret = getaddrinfo(host, NULL, &ai_hint, &ai);
+    if (ret != EOK) {
+        ai_hint.ai_flags = AI_CANONNAME | AI_NUMERICSERV;
+        ret = getaddrinfo(host, NULL, &ai_hint, &ai);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "getaddrinfo() failed (%d): %s\n", ret, gai_strerror(ret));
+            goto done;
+        } else {
+            canonname = ai->ai_canonname;
+        }
+    } else {
+        ret = getnameinfo(ai->ai_addr, ai->ai_addrlen,
+                          canonhost, NI_MAXHOST, NULL, 0, NI_NAMEREQD);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "getnameinfo() failed (%d): %s\n", ret, gai_strerror(ret));
+            goto done;
+        } else {
+            canonname = canonhost;
+        }
+    }
+    DEBUG(SSSDBG_FUNC_DATA, "Found canonical name: %s\n", canonname);
```

Comment:
> This code blocks begs to be unified/shared with
> 
> https://github.com/SSSD/sssd/blob/3caac5f7b0d1e21f9ae578f1da5324dc272aa441/src/sss_client/ssh/sss_ssh_knownhostsproxy.c#L252
> 
The code in `sss_ssh_knownshostsproxy.c` handles the port, which this code doesn't handle as it is not needed. This is what we discussed with @pbrezina earlier today.

At the same time, as `sss_ssh_knownshostsproxy.c` is obsoleted and will be removed at some time, this other code will disappear, making the sharing useless.


> Additionally I don't understand why ai_flags contain anything besides AI_NUMERICHOST/AI_CANONNAME -- AI_NUMERICSERV here and plus AI_ADDRCONFIG there?

`AI_NUMERICSERV` is a left over from previous modifications. I will remove it.
`AI_ADDRCONFIG` was removed earlier today on Pavel's request... Are you checking an old version?

Reply:
> At the same time, as `sss_ssh_knownshostsproxy.c` is obsoleted and will be removed at some time, this other code will disappear, making the sharing useless.

When? :)
https://access.redhat.com/support/policy/updates/errata#RHEL9_Planning_Guide
It probably can only be removed completely when C9S stops rebasing, not any time soon.

re. AI_ADDRCONFIG  --  I meant in sss_ssh_knownhostsproxy.c

Reply:
> > At the same time, as `sss_ssh_knownshostsproxy.c` is obsoleted and will be removed at some time, this other code will disappear, making the sharing useless.
> 
> When? :) https://access.redhat.com/support/policy/updates/errata#RHEL9_Planning_Guide It probably can only be removed completely when C9S stops rebasing, not any time soon.

Sometime. It doesn't matter when. It matters that it will happen.


> 
> re. AI_ADDRCONFIG -- I meant in sss_ssh_knownhostsproxy.c

OK. So nothing to do with this PR.